graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -0,0 +1,1560 @@
+                    train_rate=label_conf.split_rate["train"],
+                    val_rate=label_conf.split_rate["val"],
+                    test_rate=label_conf.split_rate["test"],
+                )
+            else:
+                split_rates = None
+            label_split_dicts = self._create_split_files_from_rates(
+                nodes_df, label_conf.cols[0], split_rates, split_masks_output_prefix
+            )
+            node_type_label_metadata.update(label_split_dicts)
+
+        return node_type_label_metadata
+
+    def write_edge_structure(
+        self, edge_df: DataFrame, edge_config: EdgeConfig
+    ) -> Tuple[DataFrame, Sequence[str], Sequence[str]]:
+        """Write edge structure to storage
+
+        Given a DataFrame of edges and corresponding edge configuration will
+        use the node str-to-int-id mapping to convert the node str ids to
+        ints and write the edge structure as a file with two
+        columns: "src_int_id","dst_int_id".
+
+        If `self.add_reverse_edges` is True, it it will also write the
+        reverse edge type with the src and dst columns reversed.
+
+        Parameters
+        ----------
+        edge_df : DataFrame
+            A DataFrame of edges, with columns for the source and destination node
+            types.
+        edge_config : EdgeConfig
+            The edge configuration object.
+
+        Returns
+        -------
+        Tuple[DataFrame, Sequence[str], Sequence[str]]
+            A tuple containing the original edge dataframe with ids converted to int,
+            followed by two lists of the files that were written by Spark, as S3 URIs.
+            The first list contains the original edge files, the second is the reversed
+            edge files, will be empty if `self.add_reverse_edges` is False.
+        """
+        src_col = edge_config.src_col
+        src_ntype = edge_config.src_ntype
+        dst_col = edge_config.dst_col
+        dst_ntype = edge_config.dst_ntype
+        edge_type = (
+            f"{edge_config.src_ntype}:{edge_config.get_relation_name()}:{edge_config.dst_ntype}"
+        )
+        rev_edge_type = (
+            f"{edge_config.dst_ntype}:rev-{edge_config.get_relation_name()}:{edge_config.src_ntype}"
+        )
+
+        src_node_id_mapping = (
+            self.spark.read.parquet(
+                *[
+                    os.path.join(self.output_prefix, src_path)
+                    for src_path in self.node_mapping_paths[src_ntype]
+                ]
+            )
+            .withColumnRenamed(NODE_MAPPING_INT, "src_int_id")
+            .withColumnRenamed(NODE_MAPPING_STR, "src_str_id")
+        )
+        # Join incoming edge df with mapping df to transform source str-ids to int ids
+        edge_df_with_int_src = src_node_id_mapping.join(
+            edge_df,
+            src_node_id_mapping["src_str_id"] == edge_df[src_col],
+            how="inner",
+        )
+
+        if self.enable_assertions:
```

Comment:
It's expensive yes. Because `count()` triggers a computation it forces the execution DAG to be materialized. If we don't do that we can continue applying transformations to the DF, until the point we write it do disk (or trigger another action). That allows Spark to optimize the long execution plan, vs. having to materialize the middle of the computation.